### PR TITLE
spike: move test to installCheck phase

### PR DESCRIPTION
**Pull request by blitz**
###### Motivation for this change

The test doesn't need to run in an actual VM, so there is no harm in moving it to the installCheck phase.

Relates to #143681.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/virtualization/spike/default.nix [hunk 1]
```diff
@@ -28,9 +28,24 @@ stdenv.mkDerivation rec {
...
+  installCheckInputs =
+    let
+      riscvPkgs = import ../../../.. { crossSystem = lib.systems.examples.riscv64-embedded; };
+    in
+    [ which riscvPkgs.riscv-pk riscvPkgs.hello ];
+
+  installCheckPhase = ''
+    runHook preInstallCheck
+
+    $out/bin/spike -m64 $(which pk) $(which hello) | grep -Fq "Hello, world"
```
Comment (tfc) [inline pkgs/applications/virtualization/spike/default.nix:45]: it seems simpler to not depend on `which` by simply writing 

```suggestion
    $out/bin/spike -m64 ${riscvPkgs.riscv-pk}/bin/pk ${riscvPkgs.hello}/bin/hello | grep -Fq "Hello, world"
```

or am i missing something?

Comment (blitz) [reply pkgs/applications/virtualization/spike/default.nix:45]: That's actually how it was in the NixOS integration test. I can change it back. :)

Comment (blitz) [reply pkgs/applications/virtualization/spike/default.nix]: Fixed.

File: pkgs/applications/virtualization/spike/default.nix [hunk 2]
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchgit, dtc, nixosTests, fetchpatch }:
+{ lib, stdenv, fetchgit, dtc, fetchpatch, which }:
```
Comment (SuperSandro2000) [inline pkgs/applications/virtualization/spike/default.nix:1]:
```suggestion
{ lib, stdenv, fetchgit, dtc, fetchpatch }:
```

**Comment by dramforever** [comment]
This bit:

```nix
let
  riscvPkgs = import ../../../.. { crossSystem = lib.systems.examples.riscv64-embedded; };
in
  # ...
```

Breaks completely in pure evaluation mode, because it doesn't pass `localSystem` and `builtins.currentSystem` is unavailable, rendering the spike package unusable for, e.g., flakes users unless `--impure` is turned on.

Any idea on how we should be fixing this? Depending on `pkgsCross.riscv64-embedded.{riscv-pk,hello}` is an idea, but this is the first time I've ever encountered something like this.

**Pull request merged**